rosetta-source/src/main/rosetta/event-common-func.rosetta

```diff
@@ -1834,6 +1858,384 @@ func Qualify_Shaping: <"The qualification of a shaping event from the fact that
+	inputs:
+		tradeState TradeState (1..1) <"The original trade to be for substitution of collateral.">
+		effectiveDate AdjustableOrRelativeDate (1..1) <"The date to close and open a new trade with new collateral.">
+		newCollateralPortfolio CollateralPortfolio(1..1) <"New collateral portfolio to subtitute for the original collateral.">
+		priceQuantity PriceQuantity (1..*) <"The price and quantity of the substituted product.">
+	output: 
+		instruction PrimitiveInstruction (1..1)
+
+	condition ContractualProduct: <"Only a security finance contractual product can substitute collateral.">
+		tradeState -> trade -> tradableProduct -> product -> contractualProduct exists
+
+        set instruction: <"Sets the second part of the split to be a new contract with a new asset payout.">
+                PrimitiveInstruction {
+                    quantityChange: QuantityChangeInstruction {
+                            change: priceQuantity,
+                            direction: QuantityChangeDirectionEnum -> Replace,
+                            lotIdentifier: empty
+                        },
+                    termsChange: Create_SubstitutionInstruction(
+                                tradeState -> trade -> tradableProduct -> product -> contractualProduct,
+                               effectiveDate,
+                               newCollateralPortfolio
+                            ),
+                    ...
+                }
+
+func Create_SubstitutionInstruction:<"Creates the terms change instruction that updates the payout with the new substitution payout.">
+	inputs:
+		contractualProduct ContractualProduct (1..1) <"The original contractual product to be used as the basis of the new trade.">
+		effectiveDate AdjustableOrRelativeDate (1..1) <"The effective date of the substitution.">
+		newCollateralPortfolio CollateralPortfolio(1..1) <"New collateral portfolio to substitute for the original collateral.">
+	output:
+		termsChangeInstruction TermsChangeInstruction (1..1)
+
+	set termsChangeInstruction:
+            Create_EffectiveOrTerminationDateTermChangeInstruction(contractualProduct,effectiveDate, empty)
+
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> collateral -> collateralPortfolio:
+		newCollateralPortfolio
+
+
+
+func Qualify_Substitution: <"Qualification of a collateral substitution event.">
+	[qualification BusinessEvent]
+	inputs:
+	businessEvent BusinessEvent (1..1)
+	output: is_event boolean (1..1)
+
+    alias beforeEconomicterms:
+        ExtractBeforeEconomicTerms(businessEvent)
+
+    alias openEconomicTerms:
+        ExtractOpenEconomicTerms(businessEvent)
+
+	set is_event:
+		beforeEconomicterms exists and		
+		openEconomicTerms exists and
+		openEconomicTerms -> payout <> beforeEconomicterms -> payout and
+		openEconomicTerms -> collateral <> beforeEconomicterms -> collateral and
+		openEconomicTerms -> effectiveDate = beforeEconomicterms -> terminationDate and
```

Comment:
Just want to check whether the effective date and the termination date after the substitution should be the same as the original termination date?

Otherwise I'll approve this based upon the understanding that Leo mentioned about how the remaining queries will be addressed in the next release.